### ISSUE TITLE
IME Support - Solution to long strings being truncated results in SDL_Event breaking change

### ISSUE THREAD

**Issue by zreedy** [libsdl-org/SDL#5363]
Hello,

We have been busy fixing various issues with IME support on the Windows platform in our fork. One of the issues is related to "long" strings being truncated. In some IMEs (particularly Chinese IMEs) it is common for the whole RESULTSTRING to be typed out in the IME and then sent wholesale to the client. Currently there are only 32 bytes available for text in `SDL_TextInputEvent` and `SDL_TextEditingEvent`. Because of this any sentence written that's larger than 16 characters is truncated. This makes any practical usage of these IMEs impossible for most of out Chinese and Japanese users.

Our solution so far has been to replace the char[32] buffer with a char* heap allocated buffer. This is a breaking change and in following the (apparent) standard set forward in `SDL_DropEvent` moved the responsibility of freeing this buffer to the consumer. This works in our current builds and we haven't noticed any impact on performance or memory metric.

However, I suspect that this breaking change will not be accepted upstream. I'm also unhappy with the idea of heap allocating a 2 byte buffer for the most common case which is `SDL_TextInputEvent` being issued for each key pressed on a keyboard without an IME active.

Brainstorming some solutions I can see any of the following being viable:
- `SDL_TextInputEvent` (only) gets a union for a 28 byte text buffer, and a char* pointer. With an additional flag for which is used. If the char* is used then it's the responsibility of the consumer to free the buffer. Otherwise current behavior is retained. I believe this approach will be the most backwards compatible approach.
- SDL allocates an internal heap buffer and the user must copy the text if into their memory space if they need to manipulate it. This will mean that we can simply expand the buffer as needed and avoid allocations per keystroke. This is a slight breaking change since anyone currently relying on the 32 byte buffer being mutable will see this break.

This is a serious issue that must be considered and needs to be fixed for our users. We would like to contribute our changes back up stream so we're hoping to have a discussion on what is an acceptable change here so that the PR will be accepted.

You can see our current changes here: https://github.com/YoYoGames/SDL/tree/zreedy-ime-feb22

Cheers,
Zach Reedy

**Comment by icculus** [comment]
This has been a long standing concern, so I'm putting this on the SDL3 todo list, even if we do a hint for SDL2.

**Comment by DanielGibson** [comment]
Just a quick note on:

>  I'm also unhappy with the idea of heap allocating a 2 byte buffer for the most common case which is SDL_TextInputEvent being issued for each key pressed on a keyboard without an IME active.

2 bytes wouldn't be enough for that even on most keyboards layouts that don't an IME or (much) composition, as they have many non-ASCII-chars. Both obvious ones like `Ä` or `ß` or `Ñ` or `Я` , and less obvious ones like <code>´</code> or `§`

Some kind of union for the buffer with the option for a dynamically allocated length sure sounds the way to go, though I'm not sure which of the two option (requiring the user to free the buffer vs a buffer that is only valid until the next event is received so SDL can automatically discard is) is better.
The second is less likely to create memory leaks, but OTOH it means that the user can't just copy (those) events around but needs to manually duplicate the string if it's still needed later (but won't they usually do that anyway? probably append it to an existing stringbuffer or sth)

**Comment by DanielGibson** [comment]
I just wondered, wouldn't it be possible, as a quick backwards-compatible hack, to split the string from the IME into multiple SDL_TextInput events with (UTF-8-compliant) parts of the string that fit in the existing 32char buffer?

**Comment by zreedy** [comment]
With regards to allowing the user to copy the string. Currently the string is actually duplicated on the SDL side before dispatching the event. So in theory it should be safe for a consumer to mutate the data in that pointer. I copied that behavior from `SDL_DropFile`. It might add a slight bit of overhead but it seems safer than giving the consumer a pointer to something truly internal.

> I just wondered, wouldn't it be possible, as a quick backwards-compatible hack, to split the string from the IME into multiple SDL_TextInput events with (UTF-8-compliant) parts of the string that fit in the existing 32char buffer?
Unfortunately it is not, this was one of the first things I tried. The problem is that the intention of the SDL_TEXTEDITING event is to say "here is the text as it is being reported from the IME", SDL_TEXTINPUT is similar with some additional parameters for the string length and cursor position which require knowing the full string to appropriately handle behavior on the consumer side. Ultimately it ends being ambiguous what the events mean when they're issued like this.

I don't believe that adding a flag as @slouken mentioned and unioning this data would be problematic. I will need to check to see if it's possible to retain the full 32 byte buffer in that union or if we need to make room for a 4 byte "usage" flag. Ideally we can keep the full 32 byte buffer as that would be fully backwards compatible. (Although, I'm really not sure anyone has a strict dependence on the current buffer length? If they do they really should be using the macro that defines the length rather than a hardcoded number).

**Comment by slouken** [comment]
I think there's enough room in the event structure, and if there's not, there's a compile time assert that will warn you.

**Comment by DanielGibson** [comment]
Ok, yeah, looks like SDL_TextEditingEvent can't just be split up.

Apparently SDL_Event is 56bytes (see the padding line in the union and the compile time assert afterwards).
SDL_TextEditingEvent is 20 bytes (from 5 int32s) + 32bytes (text) = 52 bytes
so there should be enough space at the end for another flag.

But then again, there's another possibility, assuming textEditingEvent.text[0] is never `\0`:

```c
typedef struct SDL_TextEditingEvent
{
    Uint32 type;                                /**< ::SDL_TEXTEDITING */
    Uint32 timestamp;                           /**< In milliseconds, populated using SDL_GetTicks() */
    Uint32 windowID;                            /**< The window with keyboard focus, if any */
    union {
        Uint8 usesInlineBuffer;
        char text[SDL_TEXTEDITINGEVENT_TEXT_SIZE];  /**< The editing text */
        struct buf {
            Uint8 _usesInlineBuffer; // same memory as usesInlineBuffer or text[0]
            char* buf; // dynamically allocated buffer
            // could also add a Uint32 bufLen, if that makes any sense
        };
    };
    
    Sint32 start;                               /**< The start cursor of selected editing text */
    Sint32 length;                              /**< The length of selected editing text */
} SDL_TextEditingEvent;
```
If `usesInlineBuffer` is 0 (`\0`), you use `textEditingEvent.dyn.buf` (that you'll have to free), otherwise you use `.text` to access the string, like before.

If the special case of `text[0] == \0` *can* happen (I never used SDL_TextEditingEvent and the documentation/tutorial isn't overly helpful, IMO; I know it doesn't make sense for SDL_TextInputEvents), that could still be handled with `dyn.buf = NULL` (this special case would need to be explicitly documented, of course).

**Comment by slouken** [comment]
I'd prefer there be an explicit flag rather than relying on someone reading the documentation carefully.

**Comment by zreedy** [comment]
I'm working on this now and as it turns out there isn't enough room in the `SDL_TextEditingEvent` without using the approach that @DanielGibson the compile-time check fails in MSVC2019.

```c
typedef struct SDL_TextEditingEvent
{
    Uint32 type;                                /**< ::SDL_TEXTEDITING */
    Uint32 timestamp;                           /**< In milliseconds, populated using SDL_GetTicks() */
    Uint32 windowID;                            /**< The window with keyboard focus, if any */
    union {                                     /**< The editing text */
        char *pText;
        char text[SDL_TEXTEDITINGEVENT_TEXT_SIZE];
    };

    Uint32 textIsPointer;                       /**< Non-zero if text is found in pText, otherwise in text buffer */
    Sint32 start;                               /**< The start cursor of selected editing text */
    Sint32 length;                              /**< The length of selected editing text */
} SDL_TextEditingEvent;
```

By my count this is 56 bytes exactly, so I'm unsure why the static check is failing.

**Comment by slouken** [comment]
In order to preserve the ABI, textIsPointer needs to come after length. I would probably make it a Uint8 and add explicit padding, e.g.
```c
    Uint8 padding1;
    Uint8 padding2;
    Uint8 padding3;
```

**Comment by slouken** [comment]
You should probably check the offset of text before and after making it a union. Some compilers align unions different from other data members. If the offset does change, that would make this an ABI breaking change and would have to go into SDL 3.

**Comment by slouken** [comment]
Another alternative would be to make a new event that just has the pointer, and send that instead of the old one, when the hint specifies that the application is expecting it.

**Comment by DanielGibson** [comment]
I guess that this is an alignment problem on 64bit architectures (I haven't checked it, but you could by printing `offsetof(SDL_TextEditingEvent, windowId)`, `offsetof(SDL_TextEditingEvent, pText)` etc).

The union comes after the windowID, i.e. after 3 32bit ints, i.e. with an offset of 12 bytes (from the beginning of the SDL_Event struct).
This doesn't matter for `char text[..]` (which has no alignment requirements), but it does for `char *pText;` which, on 64bit architectures, must be aligned to multiples of **8** bytes. I *think* that the members of a union all "start" at the same address, so the 4 byte padding (=> 12 bytes offset to 16 bytes offset) will be added before the union (and not just before `pText` within the union).

I'm not sure if this can be fixed in a portable way (i.e. without `#pragma pack()`) - even if you use sth like
```c
union {
  char text[SDL_TEXTEDITINGEVENT_TEXT_SIZE];
  struct {
    char _padding[4];
    char* textPtr;
  } buf;
};
```
that `buf` struct probably will have implicit padding between `_padding` and `textptr` and the whole struct will still require 8byte alignment.. (might be worth a try anyway, maybe there's special rules for anon structs in unions in structs or sth)

What sucks about this situation is that even the hack I suggested won't help because we can't introduce 4 bytes of padding before `text[]` without breaking the ABI

**Comment by zreedy** [comment]
I've verified that indeed the 16 byte alignment of the char* pointer appears to be the issue. Unless we can figure out some fancy union/struct packing magic we're going to need to make new events for this it seems.
This will require making a new event type for `SDL_TextInputEvent` and `SDL_TextEditingEvent` event both.
Technically, `SDL_TextInputEvent` has enough room for the flag but I think for simplicity sake it would be better to add an `SDL_TextInputEventLong`, and `SDL_TextEditingEventLong` event. Consumers can somewhat abstract away this detail by dispatching to the same handler on their end, just with a slight modification in what struct of the SDL_Event union they peek at (and free memory, as required).

These additional events will only be dispatched when the following 2 conditions are met:
- SDL_HINT_IME_SUPPORT_LONG_TEXT is SDL_TRUE
- The size of the unicode string is larger than the buffer size in bytes (32 bytes in both cases)

Does this sound like an appropriate compromise?

**Comment by slouken** [comment]
Yes, that sounds good. I'm not sure about the name *Long, but that will work for a draft PR.

**Comment by DanielGibson** [comment]
You could still only add the Long counterpart for SDL_TextEditingEvent and split up SDL_TextInputEvent into several events (as described above) - as far as I understood the splitting up is only a problem for the text **editing** event - or are those events somehow expect to correspond 1-to-1, like one set of text *editing* events that's "finished" by **exactly one** text *input* event?

Would minimize the needed changes in applications that want to support longer strings

**Comment by zreedy** [comment]
@DanielGibson I'm not entirely sure that we can split `SDL_TextInputEvent` into multiple event dispatches. I will check on my end to see if that works.
I should have a draft PR today.

Thank you both for the feedback and assistance on this so far!

**Comment by slouken** [comment]
Splitting it into multiple events is probably better, as long as it doesn't interfere with IME operation. The text is defined to be UTF-8, so splitting it in the single event delivery code path is probably fine.

**Comment by zreedy** [comment]
I realized yesterday after my post that all text is standardized to UTF-8 when issuing events. I didn't realize that `SDL_utf8strlcpy` returned bytes copied, so I overengineered the first solution.

Anyways, I've renamed "Long" to "Ext" for extended. Tidying the code up now and then I will issue a PR for the draft review.

- SDL_TextEditingExtEvent dispatched if enabled and required
- SDL_TextInputEvent is dispatched multiple times if enabled and required
